@xen-orchestra/backups/disks/MergeRemoteDisk.mjs

```diff
@@ -0,0 +1,237 @@
+                await this.#step_mergeBlocks(parentDisk, childDisk)
+            case 'cleanup':
+                await this.#step_cleanup(parentDisk, childDisk)
+                return this.#cleanup(parentDisk, childDisk)
+            default:
+                warn(`Step ${this.#state.step} is unknown`, { state: this.#state })
+        }
+        /* eslint-enable no-fallthrough */
+    }
+
+    async #writeState() {
+        try {
+            await this.#handler.writeFile(this.#statePath, JSON.stringify(this.#state), { flags: 'w' })
+        } catch (err) {
+            warn('failed to write merge state', { error: err })
+        }
+    }
+
+    async #writeStateThrottled() {
+        const delay = 10e3
+        const now = Date.now()
+        if (now - this.#lastStateWrittenAt > delay) {
+            this.#lastStateWrittenAt = now
+            await this.#writeState()
+        }
+    }
+
+    /**
+     * @param {RemoteDisk} parentDisk
+     * @param {RemoteDisk} childDisk
+     */
+    async #step_mergeBlocks(parentDisk, childDisk) {
+        const maxTableEntries = childDisk.getMaxTableEntries()
+
+        if (!this.#state) {
+            this.#state = {
+                child: { uuid: childDisk.getUuid() ?? 0 },
+                parent: { uuid: parentDisk.getUuid() ?? 0 },
+                currentBlock: 0,
+                mergedDataSize: 0,
+                step: 'mergeBlocks',
+            }
+            while (this.#state.currentBlock < maxTableEntries && !childDisk.hasBlock(this.#state.currentBlock)) {
+                ++this.#state.currentBlock
+            }
+            await this.#writeState()
+        }
+
+        await this.#mergeBlocks(parentDisk, childDisk)
+
+        await parentDisk.mergeMetadata(childDisk)
+    }
+
+    /**
+     * @param {RemoteDisk} parentDisk
+     * @param {RemoteDisk} childDisk
+     */
+    async #mergeBlocks(parentDisk, childDisk) {
+        const toMerge = []
+        for (const block of childDisk.getBlockIndexes()) {
+            if (childDisk.hasBlock(block)) toMerge.push(block)
+        }
+
+        const nBlocks = toMerge.length
+        this.#onProgress({ total: nBlocks, done: 0 })
+
+        const merging = new Set()
+        let counter = 0
+
+        await asyncEach(
+            toMerge,
+            async blockId => {
+                merging.add(blockId)
+
+                const blockSize = await parentDisk.writeBlock(blockId, (await childDisk.readBlock(blockId)).data)
+                this.#state.mergedDataSize += blockSize
+
+                this.#state.currentBlock = Math.min(...merging)
+                merging.delete(blockId)
+
+                this.#onProgress({ total: nBlocks, done: counter + 1 })
+                counter++
+                await this.#writeStateThrottled()
+            },
+            { concurrency: this.#mergeBlockConcurrency }
+        )
+
+        await this.#writeState()
+        this.#state.diskSize = await childDisk.getVirtualSize()
```

Comment:
the disk size is the size used by the disk on the file system. The virtual size is the size it would be if completly full